**Importing Libraries**

In [1]:
import numpy as np
import pandas as pd

**Loading the dataset**

In [2]:
movies=pd.read_csv('tmdb_5000_movies.csv')
credits=pd.read_csv('tmdb_5000_credits.csv')

In [3]:
movies.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

In [4]:
credits.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [5]:
credits.head(1)['crew'].values

array(['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cam

In [6]:
movies.shape

(4803, 20)

In [7]:
credits.shape

(4803, 4)

**Merge both the datasets [New dataset will have 23 col as 20 in movies 4 in credits merged on basis of title common col**

In [8]:
movies=movies.merge(credits,on='title')

In [9]:
movies.shape

(4809, 23)

In [10]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

**Col we will keep** -> genres,movie_id,keywords,title,overview,cast,crew                                                             

 **Col we will remove** -> budget,homepage,original_language[since max movies in English],original_title,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,vote_average,vote_count                                                

In [11]:
movies['original_language'].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
id       2
cs       2
ta       2
ro       2
ar       2
te       1
hu       1
xx       1
af       1
is       1
tr       1
vi       1
pl       1
nb       1
ky       1
no       1
sl       1
ps       1
el       1
Name: count, dtype: int64

In [12]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [13]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [14]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

**Check for missing data**

In [15]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [16]:
#Since 3 movies overview is missing we will drop them
movies.dropna(inplace=True)

**Checking for Duplicated data**

In [17]:
movies.duplicated().sum()

np.int64(0)

**Genre, Keywords, Cast, Crew**

In [18]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

**Covert '[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'
into ['Action','Adventure','Fantasy','Science Fiction']**

convert('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]') is not working as it is string of list we have to convert it into list => use ast.literal_eval(obj)

In [19]:
import ast
def convert(obj):
    L=[]
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [20]:
movies['genres']=movies['genres'].apply(convert)

In [21]:
movies['genres']

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [22]:
movies.iloc[0].keywords

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [23]:
movies['keywords']=movies['keywords'].apply(convert)

In [24]:
movies.iloc[0].cast

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [25]:
import ast
def convert2(obj):
    L=[]
    counter=0
    for i in ast.literal_eval(obj):
        if counter!=3:
            L.append(i['name'])
            counter+=1
        else:
            break
    return L

In [26]:
movies['cast']=movies['cast'].apply(convert2)

In [27]:
movies.iloc[0].crew

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [28]:
import ast
def convert3(obj):
    L=[]
    for i in ast.literal_eval(obj):
        if(i['job']=="Director" or i['job']=="Producer" or i['job']=="Writer" or i['job']=="Original Music Composer"):
            L.append(i['name'])
    return list(set(L)) #to remove duplicate names

In [29]:
movies['crew']=movies['crew'].apply(convert3)

In [30]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0          [James Cameron, Jon Landau, James Horner]  
1  [Jerry Bruckheimer, Peter Kohn, Chad Oman, Pat...  
2  [Barbara Broccoli, Thomas Newman, Sam Mendes, ...  
3  [Emma Thomas, Charles Roven, Christopher Nolan...  
4  [Jim Morris, Colin Wilson, Lindsey Collins, An...

Overview is in string convert it to list so that it can concatenate with other lists

In [31]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [32]:
movies['overview']=movies['overview'].apply(lambda x: x if isinstance(x,list) else x.split() if isinstance(x,str)  else []) 
#isinstance() is a built-in Python function used to check the data type of a value. isinstance("hello", str)->True

In [33]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0          [James Cameron, Jon Landau, James Horner]  
1  [Jerry Bruckheimer, Peter Kohn, Chad Oman, Pat...  
2  [Barbara Broccoli, Thomas Newman, Sam Mendes, ...  
3  [Emma Thomas, Charles Roven, Christopher Nolan...  
4  [Jim Morris, Colin Wilson, Lindsey Collins, An...

For Genres, Keywords, Cast and Crew remove all spaces between them Sam Worthington[2 entity]->SamWorthington[1 entity] for bag of words

In [34]:
movies['genres']=movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])
movies['keywords']=movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['cast']=movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew']=movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])

In [35]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]   
3      [ChristianBale, MichaelCaine, GaryOldman]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]   

                                                crew  
0             [JamesCameron, JonLandau, JamesHorner]  
1  [JerryBruckheimer, PeterKohn, ChadOman, PatSan...  
2  [BarbaraBroccoli, ThomasNewman, SamMendes, Mic...  
3  [EmmaThomas, CharlesRoven, ChristopherNolan, H...  
4  [JimMorris, ColinWilson, LindseyCollins, Andre...

**Create tags which contains overview,genres,keywords,cast,crew**

In [36]:
movies['tags']=movies['overview']+movies['genres']+movies['keywords']+movies['cast']+movies['crew']

In [37]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]   
3      [ChristianBale, MichaelCaine, GaryOldman]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]   

                                                crew  \
0             [JamesCameron, JonLandau, JamesHorner]   
1  [JerryBruckheimer, PeterKohn, ChadOman, PatSan...   
2  [BarbaraBroccoli, ThomasNewman, SamMendes, Mic...   
3  [EmmaThomas, CharlesRoven, ChristopherNolan, H...   
4  [JimMorris, ColinWilson, LindseyCollins, Andre...   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [38]:
movies.iloc[0].tags

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron',
 'JonLandau',
 'JamesHorner']

**Create new dataframe with id,title,tags remove other cols**

In [39]:
df=movies[['movie_id','title','tags']]

In [40]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [41]:
#Convert tags list to string
df.loc[:,'tags']=df['tags'].apply(lambda x: " ".join(x) if isinstance(x,list) else x) 
# .loc->label-based indexer in pandas used to select rows and columns safely.
#df.loc[row,col]->df.loc[:,tags] means label based selection on df on all rows and tags col. Select all rows of the tags column

In [42]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [43]:
df.iloc[0].tags

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron JonLandau JamesHorner'

In [44]:
#Convert tags to lower case
df.loc[:,'tags']=df['tags'].apply(lambda x: x.lower())

In [45]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

**Porter Stemmer reduce the words to their root form Ex- "running," "runner," and "ran" can all be reduced to their root form, "run."**

In [46]:
import nltk #Importing Naltural Language Toolkit

In [47]:
from nltk.stem import PorterStemmer
ps=PorterStemmer()

In [48]:
def root(text):
    y=[]
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)  #Convert list back to string

In [49]:
ps.stem('loved')

'love'

In [50]:
df.loc[:,'tags']=df['tags'].apply(root)

In [51]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a parapleg marin is dispa...  
1  captain barbossa, long believ to be dead, ha c...  
2  a cryptic messag from bond’ past send him on a...  
3  follow the death of district attorney harvey d...  
4  john carter is a war-weary, former militari ca...

In [52]:
df['tags'][0]

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron jonlandau jameshorn'

**tf-idf vectorizer**

In [53]:
from sklearn.feature_extraction.text import TfidfVectorizer
tv=TfidfVectorizer(max_features=5000,stop_words='english')
vector=tv.fit_transform(df['tags']).toarray()

In [54]:
vector.shape            
#No of movies 4806 transform it acc to 5000 features(here words)  each row represents 1 movie
#vector is the numerical representation of your movie dataset created by TF-IDF from the tags column. 
#Each movie’s text is converted into numbers so that a machine learning algorithm can understand it.
#The vectorizer selected the top 5000 most important words from all movie tags
#Each column represents one unique word (feature)
#Each movie represented as 5000 dim vector

(4806, 5000)

In [55]:
vector[0]

array([0., 0., 0., ..., 0., 0., 0.])

In [56]:
tv.get_feature_names_out().tolist()

['000',
 '007',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '17th',
 '18',
 '18th',
 '18thcenturi',
 '19',
 '1910',
 '1920',
 '1930',
 '1940',
 '1950',
 '1950s',
 '1960',
 '1960s',
 '1970',
 '1970s',
 '1980',
 '1985',
 '1990',
 '19th',
 '19thcenturi',
 '20',
 '200',
 '2009',
 '20th',
 '24',
 '25',
 '30',
 '300',
 '3d',
 '40',
 '50',
 '500',
 '60',
 '70',
 'aaron',
 'aaroneckhart',
 'aaronzigman',
 'abandon',
 'abduct',
 'abigailbreslin',
 'abil',
 'abl',
 'aboard',
 'abov',
 'abram',
 'abus',
 'academi',
 'accept',
 'access',
 'accid',
 'accident',
 'acclaim',
 'accompani',
 'accomplish',
 'account',
 'accus',
 'ace',
 'achiev',
 'act',
 'action',
 'actionhero',
 'activ',
 'activist',
 'activities',
 'actor',
 'actress',
 'actual',
 'adam',
 'adammckay',
 'adamsandl',
 'adamshankman',
 'adapt',
 'add',
 'addict',
 'adjust',
 'admir',
 'admit',
 'adolesc',
 'adopt',
 'ador',
 'adrienbrodi',
 'adult',
 'adultanim',
 'adulteri',
 'adulthood',
 'advanc',
 'adventur',
 

**Cosine Similarity**

In [57]:
from sklearn.metrics.pairwise import cosine_similarity
similarity=cosine_similarity(vector)

In [58]:
similarity.shape

(4806, 4806)

In [59]:
similarity

array([[1.        , 0.02222703, 0.02900749, ..., 0.02407089, 0.        ,
        0.        ],
       [0.02222703, 1.        , 0.01154637, ..., 0.01653154, 0.        ,
        0.00637333],
       [0.02900749, 0.01154637, 1.        , ..., 0.01498363, 0.        ,
        0.        ],
       ...,
       [0.02407089, 0.01653154, 0.01498363, ..., 1.        , 0.02013607,
        0.02908603],
       [0.        , 0.        , 0.        , ..., 0.02013607, 1.        ,
        0.0217583 ],
       [0.        , 0.00637333, 0.        , ..., 0.02908603, 0.0217583 ,
        1.        ]])

In [60]:
similarity[0]

array([1.        , 0.02222703, 0.02900749, ..., 0.02407089, 0.        ,
       0.        ])

In [61]:
similarity[0].shape

(4806,)

In [62]:
#Index of movies
df[df['title']=='Batman Begins'].index.item()

119

In [63]:
#Sort movies so that most recommended movies are shown upwards as sorted in desc order but index of the recommended movie is lost 
sorted(similarity[1].tolist(),reverse=True)

[1.0000000000000002,
 0.4116326067076084,
 0.23531022940295532,
 0.23519238574857637,
 0.18008282831375444,
 0.16986758085377235,
 0.16047394889592081,
 0.1452596578802694,
 0.1364824872376279,
 0.13399215250982568,
 0.12731962242282804,
 0.1251531599807866,
 0.1218801773450545,
 0.12163245993730397,
 0.12112739019608114,
 0.12092005743463649,
 0.115824668205681,
 0.11567544120822854,
 0.10987278153078056,
 0.10934935814483666,
 0.10856150065659131,
 0.10699828403674036,
 0.10641567013582333,
 0.10419236981346436,
 0.10351850301305156,
 0.10156284712644056,
 0.100499100137878,
 0.0996031243349817,
 0.09960260445572322,
 0.0979900233225473,
 0.09648077848585866,
 0.09586698623906595,
 0.094932042419458,
 0.0946807120069818,
 0.09369267298755722,
 0.09317270195819745,
 0.09304527222304584,
 0.09297380087137597,
 0.0928488527648254,
 0.09233767429058391,
 0.09188024719977751,
 0.0896543527404727,
 0.08711439393688462,
 0.08703331036707065,
 0.08672215154960221,
 0.08657256405419164,
 0.08

In [64]:
#To keep both index and sort 
#sorted(enumerate((similarity[0]))) -> gives enuumerate object
sorted(enumerate(similarity[1].tolist()),reverse=True) #Sorting done on basis of index

[(4805, 0.006373326553838515),
 (4804, 0.0),
 (4803, 0.016531544609434352),
 (4802, 0.0),
 (4801, 0.00954751863626633),
 (4800, 0.0),
 (4799, 0.006697190229819527),
 (4798, 0.006188315562442453),
 (4797, 0.0),
 (4796, 0.0),
 (4795, 0.023991350016159377),
 (4794, 0.010424910672096513),
 (4793, 0.0),
 (4792, 0.0),
 (4791, 0.0),
 (4790, 0.0),
 (4789, 0.0),
 (4788, 0.0),
 (4787, 0.019878192941996817),
 (4786, 0.014235566534780456),
 (4785, 0.006439029409483196),
 (4784, 0.013434160318769414),
 (4783, 0.0),
 (4782, 0.01730868933242183),
 (4781, 0.023889628830443745),
 (4780, 0.011742053096947019),
 (4779, 0.008976415279253408),
 (4778, 0.009264377579557899),
 (4777, 0.025444014860975937),
 (4776, 0.02145100298912441),
 (4775, 0.00585893697803147),
 (4774, 0.0),
 (4773, 0.0),
 (4772, 0.00468806020651499),
 (4771, 0.0),
 (4770, 0.0235887590873634),
 (4769, 0.0),
 (4768, 0.0),
 (4767, 0.0),
 (4766, 0.026977824006242764),
 (4765, 0.0),
 (4764, 0.05183554959783819),
 (4763, 0.0),
 (4762, 0.01201

In [65]:
#Sorting on basis of 2nd parameter i.e. similarity keeping index also
sorted(enumerate(similarity[1].tolist()),reverse=True,key=lambda x:x[1])

[(1, 1.0000000000000002),
 (12, 0.4116326067076084),
 (199, 0.23531022940295532),
 (17, 0.23519238574857637),
 (216, 0.18008282831375444),
 (3570, 0.16986758085377235),
 (848, 0.16047394889592081),
 (2592, 0.1452596578802694),
 (3702, 0.1364824872376279),
 (1332, 0.13399215250982568),
 (4214, 0.12731962242282804),
 (1711, 0.1251531599807866),
 (818, 0.1218801773450545),
 (2544, 0.12163245993730397),
 (109, 0.12112739019608114),
 (340, 0.12092005743463649),
 (1094, 0.115824668205681),
 (3969, 0.11567544120822854),
 (985, 0.10987278153078056),
 (3824, 0.10934935814483666),
 (75, 0.10856150065659131),
 (171, 0.10699828403674036),
 (2904, 0.10641567013582333),
 (3376, 0.10419236981346436),
 (2654, 0.10351850301305156),
 (979, 0.10156284712644056),
 (167, 0.100499100137878),
 (3699, 0.0996031243349817),
 (325, 0.09960260445572322),
 (2131, 0.0979900233225473),
 (1660, 0.09648077848585866),
 (3217, 0.09586698623906595),
 (310, 0.094932042419458),
 (187, 0.0946807120069818),
 (104, 0.09369267

In [66]:
#From the above list extract 5 most recommended movies i.e. index 12,199,17,216,3570 here
sorted(enumerate(similarity[1].tolist()),reverse=True,key=lambda x:x[1])[1:6]

[(12, 0.4116326067076084),
 (199, 0.23531022940295532),
 (17, 0.23519238574857637),
 (216, 0.18008282831375444),
 (3570, 0.16986758085377235)]

In [67]:
#Define function to find 5 most recommended movies
def recommend(movie):
    #if movie in df['title'].values:
    movie_index=df[df['title']==movie].index.item()
    dist=similarity[movie_index]
    recommend_movies=sorted(enumerate(dist.tolist()),reverse=True,key=lambda x:x[1])[1:6]
    for i in recommend_movies:
        print(df.iloc[i[0]].title)     #i is tuple contains (movie_index,similarity); iloc=integer location used to select rows and columns by index

In [68]:
recommend('Batman Begins')

The Dark Knight
The Dark Knight Rises
Batman v Superman: Dawn of Justice
Batman
Batman Returns


In [69]:
recommend('Avatar')

Aliens
Falcon Rising
Battle: Los Angeles
Aliens vs Predator: Requiem
Apollo 18


In [70]:
recommend('Iron Man')

Iron Man 2
Iron Man 3
Avengers: Age of Ultron
Ant-Man
Captain America: Civil War


**Saving trained model**

In [71]:
import pickle

In [72]:
pickle.dump(df.to_dict(),open("movies_dictionary.pkl","wb"))

In [73]:
pickle.dump(similarity,open("similarity.pkl","wb"))